File: run-length-encoding/RunLengthEncoding.fs
module RunLengthEncoding
open System

let encode input = 
    if input = "" then "" else
        let format (c, n) = (if n = 1 then "" else string n) + c.ToString()
        let counter (result, (counted, count)) c = 
            if c = counted 
            then (result, (counted, count + 1)) 
            else (result + format (counted, count), (c, 1))
        let state = Seq.fold counter ("", (input.[0], 1)) input.[1..]
        fst state + format (snd state)

let decode input = 
    let format (c:char) n = 
        if n = 0 then string c 
        else [0..n - 1] |> Seq.map (fun _ -> c) |> String.Concat
    let counter (result, count) c =
        if Char.IsNumber c 
        then (result, count * 10 + (string c |> int))
        else (result + format c count, 0)
    Seq.fold counter ("", 0) input |> fst

File: csharp/run-length-encoding/RunLengthEncoding.cs
using System;
using System.Linq;

public static class RunLengthEncoding
{
    private class CounterState
    {
        public int Count { get;set; }
        public char? Counted { get;set; }
        public string Result { get;set; } = "";

        public string AsResult() => $"{Result}{(Count == 1 ? "" : Count.ToString())}{Counted}";
    }

    public static string Encode(string input)
    {
        CounterState counter(CounterState previous, char current)
        {
            if(previous.Counted == current)
                return new CounterState 
                { 
                    Count = previous.Count + 1, 
                    Counted = current, 
                    Result = previous.Result 
                };
            var newResult = previous.Counted == null ? "" : previous.AsResult();
            return new CounterState { Count = 1, Counted = current, Result = newResult };
        }

        var processed = input.Aggregate<char, CounterState>(new CounterState(), counter);
        return processed.Count == 0 ? processed.Result : processed.AsResult();
    }

    public static string Decode(string input)
    {
        CounterState printer(CounterState previous, char current)
        {
            if(!char.IsDigit(current))
            {
                var newLength = previous.Result.Length + 
                    (previous.Count == 0 ? 1 : previous.Count);
                return new CounterState 
                { 
                    Count = 0, 
                    Result = previous.Result
                        .PadRight(newLength, current) 
                };
            }
            var number = int.Parse(current.ToString());
            return new CounterState { Count = previous.Count * 10 + number, Result = previous.Result };
        }

        var processed = input.Aggregate<char, CounterState>(new CounterState(), printer);
        return processed.Result;
    }
}
